lib/xopts.ts

```diff
@@ -0,0 +1,74 @@
+import {Signer} from 'ethers';
+import {providers} from 'ethers';
+import {ReadOnlyContracts, ReadWriteContracts} from './contracts';
+import {Addresses, Deployments} from './addresses';
+import {ExchangeRate, Currency, MonetaryAmount} from './monetary';
+
+type SignerOrProvider = Signer | providers.Provider;
+
+interface Option<From extends Currency, To extends Currency> {
+  expiry: Date;
+  from: From;
+  to: To;
+  strikePrice: ExchangeRate<From, To>;
+  currentPrice: ExchangeRate<From, To>;
+  liquidityFrom: MonetaryAmount<From>;
+}
+
+interface OptionsReadOnlyActions {
+  list(): Promise<Array<Option<Currency, Currency>>>;
+}
+
+interface OptionsReadWriteActions extends OptionsReadOnlyActions {
+  write(): Promise<void>;
+}
+
+class ContractsOptionsReadOnlyActions implements OptionsReadOnlyActions {
+  constructor(private roContracts: ReadOnlyContracts) {}
+
+  list() {
+    return Promise.resolve([]);
+  }
+}
+
+class ContractsOptionsReadWriteActions extends ContractsOptionsReadOnlyActions
+  implements OptionsReadWriteActions {
+  constructor(private contracts: ReadWriteContracts) {
+    super(contracts);
+  }
+
+  async write() {}
+}
+
+type OptionActions<T extends SignerOrProvider> = T extends Signer
+  ? OptionsReadWriteActions
+  : OptionsReadOnlyActions;
+
+export class XOpts<T extends SignerOrProvider> {
+  constructor(readonly options: OptionActions<T>) {}
+
+  static async load<T extends SignerOrProvider>(
+    provider: T,
+    addresses?: Addresses
+  ): Promise<XOpts<T>> {
+    if (addresses === undefined) {
+      addresses = Deployments.ganache; // TODO: change this to testnet/mainnet later on
```

Comment:
We need to handle both `buidlerevm` and `ganache` for local deployments. As a standard we are using buidlerevm for its speed and debugging features. ganache is only used when we want to copy the mainnet state for testing integrations.

Reply:
Greg already wrote a function to do just that so I'll update.